// @agm/drawing/esm2015/directives/drawing-manager-trigger.js
import * as tslib_1 from "tslib";
import { AgmMap } from '@agm/core';
import { Directive, Host, Input } from '@angular/core';
import { first } from 'rxjs/operators';
import { AgmDrawingManager } from './drawing-manager';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@agm/core';
let AgmDrawingManagerTrigger = class AgmDrawingManagerTrigger {
    constructor(_agmMap) {
        this._agmMap = _agmMap;
    }
    ngAfterViewInit() {
        this._agmMap.mapReady.pipe(first()).subscribe(map => this.drawingManager.setMap(map));
    }
    ngOnDestroy() {
        this._agmMap.mapReady.pipe(first()).subscribe(() => this.drawingManager.setMap(null));
    }
};
AgmDrawingManagerTrigger.ɵfac = function AgmDrawingManagerTrigger_Factory(t) { return new (t || AgmDrawingManagerTrigger)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.AgmMap, 1)); };
AgmDrawingManagerTrigger.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: AgmDrawingManagerTrigger, selectors: [["agm-map", "agmDrawingManager", ""]], inputs: { drawingManager: ["agmDrawingManager", "drawingManager"] }, exportAs: ["matDrawingManagerTrigger"] });
AgmDrawingManagerTrigger.ctorParameters = () => [
    { type: AgmMap, decorators: [{ type: Host }] }
];
tslib_1.__decorate([
    Input('agmDrawingManager'),
    tslib_1.__metadata("design:type", AgmDrawingManager)
], AgmDrawingManagerTrigger.prototype, "drawingManager", void 0);
AgmDrawingManagerTrigger = tslib_1.__decorate([ tslib_1.__param(0, Host()),
    tslib_1.__metadata("design:paramtypes", [AgmMap])
], AgmDrawingManagerTrigger);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AgmDrawingManagerTrigger, [{
        type: Directive,
        args: [{
                selector: 'agm-map[agmDrawingManager]',
                exportAs: 'matDrawingManagerTrigger'
            }]
    }], function () { return [{ type: ɵngcc1.AgmMap, decorators: [{
                type: Host
            }] }]; }, { drawingManager: [{
            type: Input,
            args: ['agmDrawingManager']
        }] }); })();
export { AgmDrawingManagerTrigger };
