// @agm/drawing/esm5/drawing.module.js
import * as tslib_1 from "tslib";
import { AgmCoreModule } from '@agm/core';
import { NgModule } from '@angular/core';
import { AgmDrawingManager } from './directives/drawing-manager';
import { AgmDrawingManagerTrigger } from './directives/drawing-manager-trigger';
import * as ɵngcc0 from '@angular/core';
var AgmDrawingModule = /** @class */ (function () {
    function AgmDrawingModule() {
    }
AgmDrawingModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: AgmDrawingModule });
AgmDrawingModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function AgmDrawingModule_Factory(t) { return new (t || AgmDrawingModule)(); }, imports: [[AgmCoreModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(AgmDrawingModule, { declarations: function () { return [AgmDrawingManager, AgmDrawingManagerTrigger]; }, imports: function () { return [AgmCoreModule]; }, exports: function () { return [AgmDrawingManager, AgmDrawingManagerTrigger]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AgmDrawingModule, [{
        type: NgModule,
        args: [{
                imports: [AgmCoreModule],
                declarations: [AgmDrawingManager, AgmDrawingManagerTrigger],
                exports: [AgmDrawingManager, AgmDrawingManagerTrigger]
            }]
    }], function () { return []; }, null); })();
    return AgmDrawingModule;
}());
export { AgmDrawingModule };
