// primeng-lts/bundles/primeng-lts-radiobutton.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/common'), require('@angular/forms')) :
    typeof define === 'function' && define.amd ? define('primeng-lts/radiobutton', ['exports', '@angular/core', '@angular/common', '@angular/forms'], factory) :
    (global = global || self, factory((global['primeng-lts'] = global['primeng-lts'] || {}, global['primeng-lts'].radiobutton = {}), global.ng.core, global.ng.common, global.ng.forms));
}(this, (function (exports, core, common, forms) { 'use strict';

    var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    };
    var RADIO_VALUE_ACCESSOR = {
        provide: forms.NG_VALUE_ACCESSOR,
        useExisting: core.forwardRef(function () { return RadioButton; }),
        multi: true
    };
    var RadioControlRegistry = /** @class */ (function () {
        function RadioControlRegistry() {
            this.accessors = [];
        }
        RadioControlRegistry.prototype.add = function (control, accessor) {
            this.accessors.push([control, accessor]);
        };
        RadioControlRegistry.prototype.remove = function (accessor) {
            this.accessors = this.accessors.filter(function (c) {
                return c[1] !== accessor;
            });
        };
        RadioControlRegistry.prototype.select = function (accessor) {
            var _this = this;
            this.accessors.forEach(function (c) {
                if (_this.isSameGroup(c, accessor) && c[1] !== accessor) {
                    c[1].writeValue(accessor.value);
                }
            });
        };
        RadioControlRegistry.prototype.isSameGroup = function (controlPair, accessor) {
            if (!controlPair[0].control) {
                return false;
            }
            return controlPair[0].control.root === accessor.control.control.root && controlPair[1].name === accessor.name;
        };
        RadioControlRegistry.ɵprov = core.ɵɵdefineInjectable({ factory: function RadioControlRegistry_Factory() { return new RadioControlRegistry(); }, token: RadioControlRegistry, providedIn: "root" });
        RadioControlRegistry = __decorate([
            core.Injectable({
                providedIn: 'root',
            })
        ], RadioControlRegistry);
        return RadioControlRegistry;
    }());
    var RadioButton = /** @class */ (function () {
        function RadioButton(cd, injector, registry) {
            this.cd = cd;
            this.injector = injector;
            this.registry = registry;
            this.onClick = new core.EventEmitter();
            this.onFocus = new core.EventEmitter();
            this.onBlur = new core.EventEmitter();
            this.onModelChange = function () { };
            this.onModelTouched = function () { };
        }
        RadioButton.prototype.ngOnInit = function () {
            this.control = this.injector.get(forms.NgControl);
            this.checkName();
            this.registry.add(this.control, this);
        };
        RadioButton.prototype.ngOnDestroy = function () {
            this.registry.remove(this);
        };
        RadioButton.prototype.handleClick = function (event, radioButton, focus) {
            event.preventDefault();
            if (this.disabled) {
                return;
            }
            this.select(event);
            if (focus) {
                radioButton.focus();
            }
        };
        RadioButton.prototype.select = function (event) {
            if (!this.disabled) {
                this.inputViewChild.nativeElement.checked = true;
                this.checked = true;
                this.onModelChange(this.value);
                this.registry.select(this);
                this.onClick.emit(event);
            }
        };
        RadioButton.prototype.writeValue = function (value) {
            this.checked = (value == this.value);
            if (this.inputViewChild && this.inputViewChild.nativeElement) {
                this.inputViewChild.nativeElement.checked = this.checked;
            }
            this.cd.markForCheck();
        };
        RadioButton.prototype.registerOnChange = function (fn) {
            this.onModelChange = fn;
        };
        RadioButton.prototype.registerOnTouched = function (fn) {
            this.onModelTouched = fn;
        };
        RadioButton.prototype.setDisabledState = function (val) {
            this.disabled = val;
        };
        RadioButton.prototype.onInputFocus = function (event) {
            this.focused = true;
            this.onFocus.emit(event);
        };
        RadioButton.prototype.onInputBlur = function (event) {
            this.focused = false;
            this.onModelTouched();
            this.onBlur.emit(event);
        };
        RadioButton.prototype.onChange = function (event) {
            this.select(event);
        };
        RadioButton.prototype.focus = function () {
            this.inputViewChild.nativeElement.focus();
        };
        RadioButton.prototype.checkName = function () {
            if (this.name && this.formControlName && this.name !== this.formControlName) {
                this.throwNameError();
            }
            if (!this.name && this.formControlName) {
                this.name = this.formControlName;
            }
        };
        RadioButton.prototype.throwNameError = function () {
            throw new Error("\n          If you define both a name and a formControlName attribute on your radio button, their values\n          must match. Ex: <p-radioButton formControlName=\"food\" name=\"food\"></p-radioButton>\n        ");
        };
        RadioButton.ctorParameters = function () { return [
            { type: core.ChangeDetectorRef },
            { type: core.Injector },
            { type: RadioControlRegistry }
        ]; };
        __decorate([
            core.Input()
        ], RadioButton.prototype, "value", void 0);
        __decorate([
            core.Input()
        ], RadioButton.prototype, "formControlName", void 0);
        __decorate([
            core.Input()
        ], RadioButton.prototype, "name", void 0);
        __decorate([
            core.Input()
        ], RadioButton.prototype, "disabled", void 0);
        __decorate([
            core.Input()
        ], RadioButton.prototype, "label", void 0);
        __decorate([
            core.Input()
        ], RadioButton.prototype, "tabindex", void 0);
        __decorate([
            core.Input()
        ], RadioButton.prototype, "inputId", void 0);
        __decorate([
            core.Input()
        ], RadioButton.prototype, "ariaLabelledBy", void 0);
        __decorate([
            core.Input()
        ], RadioButton.prototype, "ariaLabel", void 0);
        __decorate([
            core.Input()
        ], RadioButton.prototype, "style", void 0);
        __decorate([
            core.Input()
        ], RadioButton.prototype, "styleClass", void 0);
        __decorate([
            core.Input()
        ], RadioButton.prototype, "labelStyleClass", void 0);
        __decorate([
            core.Output()
        ], RadioButton.prototype, "onClick", void 0);
        __decorate([
            core.Output()
        ], RadioButton.prototype, "onFocus", void 0);
        __decorate([
            core.Output()
        ], RadioButton.prototype, "onBlur", void 0);
        __decorate([
            core.ViewChild('rb')
        ], RadioButton.prototype, "inputViewChild", void 0);
        RadioButton = __decorate([
            core.Component({
                selector: 'p-radioButton',
                template: "\n        <div [ngStyle]=\"style\" [ngClass]=\"'ui-radiobutton ui-widget'\" [class]=\"styleClass\">\n            <div class=\"ui-helper-hidden-accessible\">\n                <input #rb type=\"radio\" [attr.id]=\"inputId\" [attr.name]=\"name\" [attr.value]=\"value\" [attr.tabindex]=\"tabindex\" [attr.aria-checked]=\"checked\" [attr.aria-label]=\"ariaLabel\"\n                [attr.aria-labelledby]=\"ariaLabelledBy\" [checked]=\"checked\" (change)=\"onChange($event)\" (focus)=\"onInputFocus($event)\" (blur)=\"onInputBlur($event)\" [disabled]=\"disabled\">\n            </div>\n            <div (click)=\"handleClick($event, rb, true)\" [attr.aria-checked]=\"checked\"\n                [ngClass]=\"{'ui-radiobutton-box ui-widget ui-state-default':true,\n                'ui-state-active':rb.checked,'ui-state-disabled':disabled,'ui-state-focus':focused}\">\n                <span class=\"ui-radiobutton-icon ui-clickable\" [ngClass]=\"{'pi pi-circle-on':rb.checked}\"></span>\n            </div>\n        </div>\n        <label (click)=\"select($event)\" [class]=\"labelStyleClass\"\n            [ngClass]=\"{'ui-radiobutton-label':true, 'ui-label-active':rb.checked, 'ui-label-disabled':disabled, 'ui-label-focus':focused}\"\n            *ngIf=\"label\" [attr.for]=\"inputId\">{{label}}</label>\n    ",
                providers: [RADIO_VALUE_ACCESSOR],
                changeDetection: core.ChangeDetectionStrategy.Default
            })
        ], RadioButton);
        return RadioButton;
    }());
    var RadioButtonModule = /** @class */ (function () {
        function RadioButtonModule() {
        }
        RadioButtonModule = __decorate([
            core.NgModule({
                imports: [common.CommonModule],
                exports: [RadioButton],
                declarations: [RadioButton]
            })
        ], RadioButtonModule);
        return RadioButtonModule;
    }());

    exports.RADIO_VALUE_ACCESSOR = RADIO_VALUE_ACCESSOR;
    exports.RadioButton = RadioButton;
    exports.RadioButtonModule = RadioButtonModule;
    exports.RadioControlRegistry = RadioControlRegistry;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
